### PR TITLE
[STLExtras] Add out-of-line definition of friend operator== for C++20

### PR DESCRIPTION
**Pull request by usx95**
The last attempt at https://github.com/llvm/llvm-project/pull/72220 was reverted by https://github.com/llvm/llvm-project/commit/94d6699bf5eeb5aa4c50d1d90f8bf69b79201ceb because it breaks C++20 build in clang-17 and before.

This is a workaround of https://github.com/llvm/llvm-project/issues/70210 and unblocks https://github.com/llvm/llvm-project/pull/72213 which rectifies rewriting template operator and thus introduces new breakages.

Moving the function definition out of the class makes clang find a matching `operator!=` for the `operator==`. This makes clang not rewrite the `operator==` with reversed args. Hence, the ambiguity is resolved.

The final plan, when https://github.com/llvm/llvm-project/issues/70210 is fixed, is to move these back to inline definition or even convert to a member template operator. This should not be urgent and could even wait for a major clang release including https://github.com/llvm/llvm-project/pull/72213

### CODE REVIEW

File: llvm/include/llvm/ADT/STLExtras.h [hunk 1]
```diff
@@ -1291,16 +1291,19 @@ class indexed_accessor_range_base {
...
-  }
-  template <typename OtherT>
-  friend bool operator!=(const indexed_accessor_range_base &lhs,
-                         const OtherT &rhs) {
-    return !(lhs == rhs);
-  }
+  // FIXME: Make me a member function instead of friend when it works in C++20.
+  template <typename OtherT, typename DerivedT2, typename BaseT2, typename T2,
+            typename PointerT2, typename ReferenceT2>
+  friend bool
```
Comment (ilya-biryukov) [inline llvm/include/llvm/ADT/STLExtras.h:1297]: We do not need to declare this as a friend, the implementation only uses the public methods.
Could we remove the friend declarations completely and only keep the out-of-line functions?

**Review by ilya-biryukov — COMMENTED** [review]
I will double-check the builds and send another update.
(This time actually building the offending targets, sorry for a blunder yesterday)

**Comment by usx95** [comment]
This would need a cherry-pick for issue https://github.com/llvm/llvm-project/issues/68901 to get into 17.0.6. After that we should be able to submit this and upgrade C++20 bot to 17.0.6.

**Comment by usx95** [comment]
Update: I verified that this change works fine with clang-17.0.6 and fails with clang-17.0.5 in C++20.

**Review by cor3ntin — APPROVED** [review]
LGTM

**Comment by usx95** [comment]
I am awaiting C++20 build bots to be updated to 17.0.6.

FYI: @cor3ntin @AaronBallman with this change, clang 17.0.6 would be the minimum version to build LLVM in C++20 after this change.
Would you like me to announce it somewhere or update some docs to reflect this ?

**Comment by cor3ntin** [comment]
@usx95 stupid question, how does that affects llvm when built with GCC/MSVC?
If we do require a too recent version of compiler to build in C++20 mode it will make it harder to move llvm to C++20.

**Comment by AaronBallman** [comment]
> I am awaiting C++20 build bots to be updated to 17.0.6. (cc: @ilya-biryukov)
> 
> FYI: @cor3ntin @AaronBallman with this change, clang 17.0.6 would be the minimum version to build LLVM in C++20 after this change. Would you like me to announce it somewhere or update some docs to reflect this ?

It should be announced in the release notes as well as the Announcements channel so people are aware. That said, I am still concerned with bumping the minimum version up that much. Clang 5.0 (Sep 2017) to Clang 17.0 (Nov 2023) is quite the jump. This might not be a bad thing to float via an RFC in Discourse before announcing it as our final decision just to make sure we get community buy-in. I don't think it will be contentious (there's not much to be done; when we want to switch to C++20 for the code base, we're going to need to make a significant leap anyway), but best to make sure this won't be disruptive to anyone in ways that we can address.

**Comment by usx95** [comment]
Clang 17.0.6 fixed https://github.com/llvm/llvm-project/issues/68901. This issue was not present in GCC or MSVC [godbolt](https://godbolt.org/z/YEhK37Mqj).

I have added release notes. I will also file the RFC and wait for comments.

**Comment by usx95** [comment]
Sent out an [RFC](https://discourse.llvm.org/t/rfc-clang-17-0-6-would-be-minimum-version-to-build-llvm-in-c-20/75345).

**Comment by usx95** [comment]
Build bot has been updated, and the RFC has no pushback. I will land this now.

**Pull request merged**